### PR TITLE
BO Migration Contact > Stores - Add command, command handler and behat tests for bulk status toggle

### PR DESCRIPTION
**Pull request by LouisAUTHIE**
| Questions         | Answers
| ----------------- | -------------------------------------------------------
| Branch?           | develop
| Description?      | BO Migration Contact > Stores - Add command, command handler and behat tests for bulk status toggle
| Type?             | improvement
| Category?         | BO
| BC breaks?        | no
| Deprecations?     | no
| Fixed ticket?     | N/A
| How to test?      | Run behat test via :  ./vendor/bin/behat -c tests/Integration/Behaviour/behat.yml -s store
| Possible impacts? | Nothing

---
This change is [Reviewable](https://reviewable.io/reviews/prestashop/prestashop/27853)

### CODE REVIEW

File: tests/Integration/Behaviour/Features/Scenario/Store/store.feature
Comment (Progi1984) [inline tests/Integration/Behaviour/Features/Scenario/Store/store.feature]:
```suggestion
    When I disable multiple stores "StorePau, StoreSerresCastet" using bulk action
    Then stores "StorePau, StoreSerresCastet" should be disabled
    When I enable multiple stores "StorePau, StoreSerresCastet" using bulk action
    Then stores "StorePau, StoreSerresCastet" should be enabled
```

File: classes/Store.php
Comment (Progi1984) [inline classes/Store.php]:
```suggestion
    /** @var string|array<string> Name */
```

File: classes/Store.php
Comment (Progi1984) [inline classes/Store.php]:
```suggestion
    /** @var string|array<string> Address first line */
```

File: classes/Store.php
Comment (Progi1984) [inline classes/Store.php]:
```suggestion
    /** @var string|array<string> Address second line (optional) */
```

File: classes/Store.php
Comment (Progi1984) [inline classes/Store.php]:
```suggestion
    /** @var string|array<string> Note */
```

**Review by Progi1984 — DISMISSED** [review]

File: tests/Integration/Behaviour/Features/Context/Domain/StoreFeatureContext.php
Comment (saulaski) [inline tests/Integration/Behaviour/Features/Context/Domain/StoreFeatureContext.php]:
```suggestion
    public function assertSoreStatus(string $storeReference, string $expectedStatus): void
```

File: tests/Integration/Behaviour/Features/Scenario/Store/store.feature
Comment (saulaski) [inline tests/Integration/Behaviour/Features/Scenario/Store/store.feature]: May I suggest adding 3rd store excluded from bulk actions ? just to make sure this store doesn't get toggled by a bug later.

File: tests/Integration/Behaviour/behat.yml
Comment (saulaski) [inline tests/Integration/Behaviour/behat.yml]: Indentation here doesn't match the entire file

**Review by saulaski — CHANGES_REQUESTED** [review]

File: tests/Integration/Behaviour/Features/Context/Domain/StoreFeatureContext.php
Comment (matthieu-rolland) [reply tests/Integration/Behaviour/Features/Context/Domain/StoreFeatureContext.php]: you mean `assertStoreStatus`? :D

Comment (saulaski) [reply tests/Integration/Behaviour/Features/Context/Domain/StoreFeatureContext.php]: Yes indeed, sorry for the typo :)

File: tests/Integration/Behaviour/behat.yml
Comment (WebHelpersPau) [reply tests/Integration/Behaviour/behat.yml]: Indentation corrected for state and store

File: tests/Integration/Behaviour/Features/Scenario/Store/store.feature
Comment (WebHelpersPau) [reply tests/Integration/Behaviour/Features/Scenario/Store/store.feature]: Another shop added for test and test done to check that the status doesn't change

File: tests/Integration/Behaviour/Features/Context/Domain/StoreFeatureContext.php
Comment (WebHelpersPau) [reply tests/Integration/Behaviour/Features/Context/Domain/StoreFeatureContext.php]: Function name changed, no worries @saulaski ;)

**Review by Progi1984 — DISMISSED** [review]

**Review by Progi1984 — DISMISSED** [review]

**Review by saulaski — DISMISSED** [review]

**Review by matthieu-rolland — DISMISSED** [review]

**Comment by matthieu-rolland** [comment]
Since there is no QA on this PR, I would like another approve before merging 👍

**Comment by WebHelpersPau** [comment]
Thanks @matthieu-rolland , @Progi1984 , @saulaski for the quick and efficient review. Nice way to learn a vertical migration

**Comment by WebHelpersPau** [comment]
@jolelievre just a ping, if you'd like to follow my journey through this SF migration ! ;)

File: src/Core/Domain/Store/Command/BulkToggleStoreStatusCommand.php
Comment (jolelievre) [inline src/Core/Domain/Store/Command/BulkToggleStoreStatusCommand.php]:
```suggestion
class BulkUpdateStoreStatusCommand
```

Comment (jolelievre) [reply src/Core/Domain/Store/Command/BulkToggleStoreStatusCommand.php]: Unlike the `ToggleStoreStatusCommand` which inverts the store's status, this one has the expected status as a parameter So it's not exactly a toggle but rather an update of the status. So I suggest renaming this command to reflect more its action.

File: tests/Integration/Behaviour/Features/Context/StoreFeatureContext.php
Comment (jolelievre) [inline tests/Integration/Behaviour/Features/Context/StoreFeatureContext.php:57]: This one should be replaced with an actual CQRS command in another PR

File: tests/Integration/Behaviour/Features/Context/Domain/StoreFeatureContext.php
Comment (jolelievre) [inline tests/Integration/Behaviour/Features/Context/Domain/StoreFeatureContext.php]: You don't need this part in this context, probably a copy from another class?

File: tests/Integration/Behaviour/Features/Context/StoreFeatureContext.php
Comment (jolelievre) [inline tests/Integration/Behaviour/Features/Context/StoreFeatureContext.php:44]: We already have a `AbstractDomainFeatureContext::getDefaultLangId` helper function to do this, however you can't extend it here since this is not a "Domain context" so it's fine for


Just keep in mind when you create the `AddStoreCommand` (if you planned on doing it after), that you will need to migrate this behat endpoint into the store domain context and at this point you'll be able to use `$this->getDefaultLangId()`

That said, you will probably need to handle localized values so actually you won't even need this info anymore, the data from feature files will have to be a localized array

**Review by jolelievre — CHANGES_REQUESTED** [review]
Sorry to block the PR, but I just want to clean a few things before this one is merged

@LouisAUTHIE very nice work, nothing to say about implementation only an improvement of the command naming. The rest is comments for future PRs and some code that needs to be cleaned

Thank you!

**Comment by LouisAUTHIE** [comment]
Thanks for the review @jolelievre , changes requested are coming up very soon

**Comment by LouisAUTHIE** [comment]
Update done considering the review of @jolelievre. Rebase done

File: src/PrestaShopBundle/Resources/config/services/adapter/store.yml
Comment (jolelievre) [inline src/PrestaShopBundle/Resources/config/services/adapter/store.yml:19]:
```suggestion

  prestashop.adapter.store.command_handler.bulk_update_store_status_handler:
```

**Review by jolelievre — COMMENTED** [review]
Thanks @LouisAUTHIE 
One last comment and it's all good!

File: src/PrestaShopBundle/Resources/config/services/adapter/store.yml
Comment (LouisAUTHIE) [reply src/PrestaShopBundle/Resources/config/services/adapter/store.yml:19]: Sorry, I forgot this one! Thanks!

**Review by Progi1984 — APPROVED** [review]

**Review by matthieu-rolland — APPROVED** [review]

**Review by jolelievre — APPROVED** [review]
Thank you @LouisAUTHIE

**Pull request merged**
